linfa-clustering/src/gaussian_mixture/algorithm.rs

```diff
@@ -0,0 +1,494 @@
+/// that is maximising the probability that the dataset is drawn from our mixture of normal distributions.
+///
+/// After an initialization step which can be either from random distribution or from the result
+/// of the [KMeans](struct.KMeans.html) algorithm (which is the default value of the `init_method` parameter).
+/// The core EM iterative algorithm for Gaussian Mixture is a fixed-point two-step algorithm:
+///
+/// 1. Expectation step: compute the expectation of the likelihood of the current gaussian mixture model wrt the dataset.
+/// 2. Maximization step: update the gaussian parameters (weigths, means and covariances) to maximize the likelihood.
+///
+/// We stop iterating when there is no significant gaussian parameters change (controlled by the `tolerance` parameter) or
+/// if we reach a max number of iterations (controlled by `max_n_iterations` parameter)
+/// As the initialization of the algorithm is subject to randomness, several initializations are performed (controlled by
+/// the `n_init` parameter).   
+///
+/// ## Tutorial
+///
+/// Let's do a walkthrough of a training-predict-save example.
+///
+/// ```rust
+/// extern crate openblas_src;
+/// use linfa::Dataset;
+/// use linfa::traits::{Fit, Predict};
+/// use linfa_clustering::{GmmHyperParams, GaussianMixtureModel, generate_blobs};
+/// use ndarray::{Axis, array, s, Zip};
+/// use ndarray_rand::rand::SeedableRng;
+/// use rand_isaac::Isaac64Rng;
+/// use approx::assert_abs_diff_eq;
+///
+/// let mut rng = Isaac64Rng::seed_from_u64(42);
+/// let expected_centroids = array![[0., 1.], [-10., 20.], [-1., 10.]];
+/// let n = 200;
+///
+/// // We generate a dataset from points normally distributed around some distant centroids.  
+/// let dataset = Dataset::from(generate_blobs(n, &expected_centroids, &mut rng));
+///
+/// // Our GMM is expected to have a number of clusters equals the number of centroids
+/// // used to generate the dataset
+/// let n_clusters = expected_centroids.len_of(Axis(0));
+///
+/// // We fit the model from the dataset setting some options
+/// let gmm = GaussianMixtureModel::params_with_rng(n_clusters, rng)
+///             .n_init(10)
+///             .tolerance(1e-4)
+///             .build()
+///             .fit(&dataset);
+///
+/// // We can get predicted centroids (ie means of learnt gaussian distributions) from the model
+/// let gmm_centroids = gmm.centroids();
+///
+/// // We can check that centroids used to generate test dataset are close to GMM centroids
+/// let memberships = gmm.predict(&expected_centroids);
+/// for (i, expected_c) in expected_centroids.outer_iter().enumerate() {
+/// let closest_c = gmm_centroids.index_axis(Axis(0), memberships[i]);
+/// Zip::from(&closest_c)
+///     .and(&expected_c)
+///     .apply(|a, b| assert_abs_diff_eq!(a, b, epsilon = 1.))
+/// }
+/// ```
+pub struct GaussianMixtureModel<F: Float> {
+    covar_type: GmmCovarType,
+    weights: Array1<F>,
```

Comment:
there is also a weight field in datasets, perhaps you can reuse it here?

Reply:
You mean to allow the initialization of the weights by the user when <code>fit</code> is called?

Reply:
I misinterpreted the field as a weight to distribute different significance for each point, but that's the weight per distribution